conans/test/functional/conanfile/conan_data_test.py

```diff
@@ -0,0 +1,135 @@
+import os
+import unittest
+
+from bottle import static_file
+from nose.plugins.attrib import attr
+
+from conans.model.ref import ConanFileReference
+from conans.test.utils.test_files import tgz_with_contents
+from conans.test.utils.tools import TestClient, StoppableThreadBottle
+
+
+class ConanDataTest(unittest.TestCase):
+
+    def conan_exports_kept_test(self):
+        client = TestClient()
+        conanfile = """from conans import ConanFile
+
+class Lib(ConanFile):
+    exports = "myfile.txt"
+"""
+        client.save({"conanfile.py": conanfile,
+                     "myfile.txt": "bar",
+                     "conandata.yml": """
+foo:
+  bar: "as"                   
+"""})
+        ref = ConanFileReference.loads("Lib/0.1@user/testing")
+        client.run("export . {}".format(ref))
+        export_folder = client.cache.package_layout(ref).export()
+        self.assertTrue(os.path.exists(os.path.join(export_folder, "conandata.yml")))
+        self.assertTrue(os.path.exists(os.path.join(export_folder, "myfile.txt")))
+
+    def conan_data_everywhere_test(self):
+        client = TestClient()
+        conanfile = """from conans import ConanFile
+
+class Lib(ConanFile):
+
+    def _assert_data(self):
+        assert(self.conan_data["sources"]["all"]["url"] == "the url")
+        assert(self.conan_data["sources"]["all"]["other"] == "field")
+        self.output.info("My URL: {}".format(self.conan_data["sources"]["all"]["url"]))
+
+    def configure(self):
+        self._assert_data()
+
+    def config_options(self):
+        self._assert_data()
+
+    def source(self):
+        self._assert_data()
+
+    def build(self):
+        self._assert_data()
+        
+    def package(self):
+        self._assert_data()
+
+    def package_info(self):
+        self._assert_data()
+"""
+        client.save({"conanfile.py": conanfile,
+                     "conandata.yml": """
+sources:
+  all:
+    url: "the url"
+    other: "field"                   
+"""})
+        ref = ConanFileReference.loads("Lib/0.1@user/testing")
+        client.run("create . {}".format(ref))
+        self.assertIn("File 'conandata.yml' found. Exporting it...", client.out)
+        self.assertIn("My URL:", client.out)
+        export_folder = client.cache.package_layout(ref).export()
+        self.assertTrue(os.path.exists(os.path.join(export_folder, "conandata.yml")))
+
+        # Transitive loaded?
+        client.save({"conanfile.txt": "[requires]\n{}".format(ref)}, clean_first=True)
+        client.run("install . ")
+        self.assertIn("My URL:", client.out)
+        client.run("install . --build")
+        self.assertIn("My URL:", client.out)
+
+    @attr("slow")
+    def conan_data_as_source_test(self):
```

Comment:
I'd say that launching a server to really download that file is overkill for the scope of this feature, it doesn't deserve such an integration test, I would be happy without it.

Reply:
I agree but as it is exactly the thing we are planning for `the wheel of pain` I think it is no so bad to have it.